### DISCUSSION TITLE
From FIT of Siril to TIF (with color)

### DISCUSSION THREAD

**Discussion by FAb7D** [ImageMagick/ImageMagick#6053] [Digital Image Processing]
I'm trying to transform the photos in FIT format produced by Siril (in color) into color TIF with an Imagemagick command, but I always get only black and white photos (grayscale). What is the correct command to perform this conversion?
Thanks

**Comment by dlemstra** [comment]
Can you share your image?

**Reply by fmw42** [reply]
Share your command also.

**Comment by FAb7D** [comment]
I try with https://wetransfer.com/ (I hope work): [https://we.tl/t-fOmMIFrwtd](https://we.tl/t-fOmMIFrwtd) it's a very large file so I compressed it a bit like a gz (136 Mb). The image is very dark because it is a photo of a comet (part of a stacking), to see something you have to increase the exposure, it has a greenish color. When I open it with Gimp a window appears asking if the image composition is none or NAXIS=3 ....  Only with this last choice the file fit is open like a color photo.

My commands:
convert prova.fit -set colorspace RGB prova.tif 
convert prova.fit -profile ICC -set colorspace RGB prova.tif
and if I try this...:
magick display r_r_pp_light_00039.fit
the image is grayscale

Let me know if the link doesn't work: this is my first time using this service(wetransfer..).

**Comment by urban-warrior** [comment]
Perhaps the image has three separate channels, one for red, green, and blue.  Combine them into an RGB with this command:

```
magick prova.fit -combine  prova.tif
```

**Reply by FAb7D** [reply]
![Screenshot from 2023-02-11 20:15:00](https://user-images.githubusercontent.com/104997294/218277015-a269ea06-c0a5-43cf-9ebf-e341ebdd559e.png)

Done (`magick prova.fit -combine  prova.tif`), but the results is "red-scale", without green and blue.

I attach the photo of the gimp window to open the fit.

Opening one of these fit files with gedit, this is what it reads at the beginning:
"
SIMPLE  =                    T / file does conform to FITS standard             BITPIX  =                  -32 / number of bits per data pixel                  NAXIS   =                    3 / number of data axes                            NAXIS1  =                 5488 / length of data axis 1                          NAXIS2  =                 3664 / length of data axis 2                          NAXIS3  =                    3 / length of data axis 3                          EXTEND  =                    T / FITS dataset may contain extensions            COMMENT   FITS (Flexible Image Transport System) format is defined in 'AstronomyCOMMENT   and Astrophysics', volume 376, page 359; bibcode: 2001A&A...376..359H BZERO   =                   0. / offset data range to that of unsigned short    BSCALE  =                   1. / default scaling factor                         ROWORDER= 'BOTTOM-UP'          / Order of the rows in image array               INSTRUME= 'Pentax K-S2'        / instrument name                                DATE    = '2023-02-08T16:45:04' / UTC date that FITS file was created           DATE-OBS= '2023-01-30T00:43:19' / YYYY-MM-DDThh:mm:ss observation start, UT     EXPTIME =      10.039999961853 / Exposure time [s]                              XPIXSZ  =                  4.3 / X pixel size microns                           YPIXSZ  =                  4.3 / Y pixel size microns                           XBINNING=                    1 / Camera binning mode                            YBINNING=                    1 / Camera binning mode                            FOCALLEN=                 180. / Camera focal length                            CCD-TEMP=                   0. / CCD temp in C                                  APERTURE=     5.59999990463257 / Aperture of the instrument                     ISOSPEED=                6400. / ISO camera setting                             PROGRAM = 'Siril v1.1.0'       / Software that created this HDU                 HISTORY Calibrated with a master dark                                           HISTORY Calibrated with a master flat, normalization of 0.005                   HISTORY Cosmetic correction of 0 cold pixels and 0 hot pixels                   END "

**Comment by fmw42** [comment]
Imagemagick is seeing 3 pages/layers. The first has data between 0 and 65535, but the second and third seem to be exactly 0 (totally black).  When I open this in GraphicConverter on my Mac, I get (-resize 20%)

![r_r_pp_light_00039_small](https://user-images.githubusercontent.com/13458218/218278946-c5e6fc59-9b7d-4370-915e-1d1caea09cfb.png)

So there is proper data in each layer, but Imagemagick is not reading the FITS file properly.

**Reply by fmw42** [reply]
My error. I accidentally used IM 6 rather than IM 7, so got clipped data in identify -verbose.

However, when using IM 7 with auto-level each channel, I get a very cyan image. 

 `magick r_r_pp_light_00039.fit -combine -channel rgb -auto-level x.tiff`

and doing it with the channels together, I still get a black image

 `magick r_r_pp_light_00039.fit -combine  -auto-level x.tiff`

So there is likely something in the meta data to show how to scale the layers when combining.

**Comment by urban-warrior** [comment]
Channels 2 & 3 have a max less than zero.  ImageMagick does not automatically scale so its up to the user to decide if they want to scale a channel and by how much.  Try `identify -verbose` to see the statistics of each plane in the image.   See https://siril.org/tutorials/rgb_composition/ where is says "It can be tricky to compose a color image from images made with various filters."

**Reply by FAb7D** [reply]
```
$ magick identify -verbose r_pp_light_00039.fit
Image:
  Filename: r_pp_light_00039.fit
  Format: FITS (Flexible Image Transport System)
  Class: DirectClass
  Geometry: 5488x3664+0+0
  Units: Undefined
  Colorspace: Gray
  Type: Grayscale
  Base type: Undefined
  Endianness: MSB
  Depth: 32/16-bit
  Channel depth:
    Gray: 16-bit
  Channel statistics:
    Pixels: 20108032
    Gray:
      min: -3.03312e-11  (-4.62825e-16)
      max: 65535 (1)
      mean: 1149.79 (0.0175446)
      median: 1096.88 (0.0167374)
      standard deviation: 189.109 (0.00288562)
      kurtosis: 35001.3
      skewness: 141.094
      entropy: 0.552001
  Rendering intent: Undefined
  Gamma: 0.454545
  Matte color: grey74
  Background color: white
  Border color: srgb(223,223,223)
  Transparent color: none
  Interlace: None
  Intensity: Undefined
  Compose: Over
  Page geometry: 5488x3664+0+0
  Dispose: Undefined
  Iterations: 0
  Scene: 0 of 3
  Compression: Undefined
  Orientation: Undefined
  Properties:
    comment:   FITS (Flexible Image Transport System) format is defined in 'Astronomy  and Astrophysics', volume 376, page 359; bibcode: 2001A&A...376..359H 
    date:create: 2023-02-04T22:20:03+00:00
    date:modify: 2023-02-04T22:20:03+00:00
    date:timestamp: 2023-02-11T20:49:15+00:00
    fits:aperture: 5.59999990463257 / Aperture of the instrument                     
    fits:bitpix: -32 / number of bits per data pixel                  
    fits:bscale: 1. / default scaling factor                         
    fits:bzero: 0. / offset data range to that of unsigned short    
    fits:ccd-temp: 0. / CCD temp in C                                  
    fits:comment:   and Astrophysics', volume 376, page 359; bibcode: 2001A&A...376..359H 
    fits:date: '2023-02-04T22:20:02' / UTC date that FITS file was created           
    fits:date-obs: '2023-01-30T00:53:59' / YYYY-MM-DDThh:mm:ss observation start, UT     
    fits:exptime: 10.039999961853 / Exposure time [s]                              
    fits:extend: T / FITS dataset may contain extensions            
    fits:focallen: 180. / Camera focal length                            
    fits:history: Cosmetic correction of 0 cold pixels and 0 hot pixels                   
    fits:instrume: 'Pentax K-S2'        / instrument name                                
    fits:isospeed: 6400. / ISO camera setting                             
    fits:naxis: 3 / number of data axes                            
    fits:naxis1: 5488 / length of data axis 1                          
    fits:naxis2: 3664 / length of data axis 2                          
    fits:naxis3: 3 / length of data axis 3                          
    fits:program: 'Siril v1.1.0'       / Software that created this HDU                 
    fits:roworder: 'BOTTOM-UP'          / Order of the rows in image array               
    fits:simple: T / file does conform to FITS standard             
    fits:xbinning: 1 / Camera binning mode                            
    fits:xpixsz: 4.3 / X pixel size microns                           
    fits:ybinning: 1 / Camera binning mode                            
    fits:ypixsz: 4.3 / Y pixel size microns                           
    signature: 5f1fbcc6757908df8f216259afaa843c13e78c1d3797db1546d59821677bd9c2
  Artifacts:
    verbose: true
  Tainted: False
  Filesize: 0B
  Number pixels: 20.108M
  Pixel cache type: Memory
  Pixels per second: 19.123MP
  User time: 2.070u
  Elapsed time: 0:02.051
  Version: ImageMagick 7.1.0-48 Q16-HDRI x86_64 c1ca24765:20220910 https://imagemagick.org
Image:
  Filename: r_pp_light_00039.fit
  Format: FITS (Flexible Image Transport System)
  Class: DirectClass
  Geometry: 5488x3664+0+0
  Units: Undefined
  Colorspace: Gray
  Type: Grayscale
  Base type: Undefined
  Endianness: MSB
  Depth: 32/16-bit
  Channel depth:
    Gray: 16-bit
  Channel statistics:
    Pixels: 20108032
    Gray:
      min: -7157.76  (-0.10922)
      max: -27.7555 (-0.000423521)
      mean: -148.59 (-0.00226733)
      median: -147.148 (-0.00224533)
      standard deviation: 32.3115 (0.000493041)
      kurtosis: 14490.5
      skewness: -92.2533
      entropy: 0
  Rendering intent: Undefined
  Gamma: 0.454545
  Matte color: grey74
  Background color: white
  Border color: srgb(223,223,223)
  Transparent color: none
  Interlace: None
  Intensity: Undefined
  Compose: Over
  Page geometry: 5488x3664+0+0
  Dispose: Undefined
  Iterations: 0
  Scene: 1 of 3
  Compression: Undefined
  Orientation: Undefined
  Properties:
    date:create: 2023-02-04T22:20:03+00:00
    date:modify: 2023-02-04T22:20:03+00:00
    date:timestamp: 2023-02-11T20:49:15+00:00
    signature: eb7717adfe071d361e6152f63722918232918e350415379428f2d9e9284edc54
  Artifacts:
    verbose: true
  Tainted: False
  Filesize: 0B
  Number pixels: 20.108M
  Pixel cache type: Memory
  Pixels per second: 14.6994MP
  User time: 2.130u
  Elapsed time: 0:02.367
  Version: ImageMagick 7.1.0-48 Q16-HDRI x86_64 c1ca24765:20220910 https://imagemagick.org
Image:
  Filename: r_pp_light_00039.fit
  Format: FITS (Flexible Image Transport System)
  Class: DirectClass
  Geometry: 5488x3664+0+0
  Units: Undefined
  Colorspace: Gray
  Type: Grayscale
  Base type: Undefined
  Endianness: MSB
  Depth: 32/16-bit
  Channel depth:
    Gray: 16-bit
  Channel statistics:
    Pixels: 20108032
    Gray:
      min: -6991.6  (-0.106685)
      max: 20.0775 (0.000306363)
      mean: -141.852 (-0.00216452)
      median: -136.961 (-0.0020899)
      standard deviation: 25.7437 (0.000392823)
      kurtosis: 21534
      skewness: -111.222
      entropy: 1.27829e-06
  Rendering intent: Undefined
  Gamma: 0.454545
  Matte color: grey74
  Background color: white
  Border color: srgb(223,223,223)
  Transparent color: none
  Interlace: None
  Intensity: Undefined
  Compose: Over
  Page geometry: 5488x3664+0+0
  Dispose: Undefined
  Iterations: 0
  Scene: 2 of 3
  Compression: Undefined
  Orientation: Undefined
  Properties:
    date:create: 2023-02-04T22:20:03+00:00
    date:modify: 2023-02-04T22:20:03+00:00
    date:timestamp: 2023-02-11T20:49:16+00:00
    signature: c3d58f1994f00f832a9088c4f57716bb238153616978673c3ed85061f5d757b7
  Artifacts:
    verbose: true
  Tainted: False
  Filesize: 230.122MiB
  Number pixels: 20.108M
  Pixel cache type: Memory
  Pixels per second: 10.661MP
  User time: 2.460u
  Elapsed time: 0:02.886
  Version: ImageMagick 7.1.0-48 Q16-HDRI x86_64 c1ca24765:20220910 https://imagemagick.org
```

**Reply by FAb7D** [reply]
However the photo is taken by a reflex directly in RGB raw (no filters just a telephoto lens). 
![Screenshot Siril](https://user-images.githubusercontent.com/104997294/218283155-0e8e4ab6-66bd-4ca8-b945-f994b154ad39.png)
NB the image is visualized like this because the visualization is "autostrech", in reality it is linear and much darker.

**Comment by FAb7D** [comment]
Now I tried saving from Siril the same image as a Tiff (NB:32 bit), the image is correct (very dark) and the output of `magick identify -verbose` is:
```
Image:
  Filename: provadasiril.tif
  Format: TIFF (Tagged Image File Format)
  Mime type: image/tiff
  Class: DirectClass
  Geometry: 5488x3664+0+0
  Units: PixelsPerInch
  Colorspace: sRGB
  Type: TrueColor
  Endianness: LSB
  Depth: 32/16-bit
  Channel depth:
    Red: 16-bit
    Green: 16-bit
    Blue: 16-bit
  Channel statistics:
    Pixels: 20108032
    Red:
      min: -97.9251  (-0.00149424)
      max: 7686.34 (0.117286)
      mean: 6.71508 (0.000102465)
      median: 1.69153 (2.58111e-05)
      standard deviation: 21.7435 (0.000331784)
      kurtosis: 40329.9
      skewness: 156.508
      entropy: 0.349347
    Green:
      min: -96.9943  (-0.00148004)
      max: 7029.03 (0.107256)
      mean: 17.1197 (0.000261229)
      median: 17.0246 (0.000259779)
      standard deviation: 31.5567 (0.000481524)
      kurtosis: 15287.3
      skewness: 95.8119
      entropy: 0.40985
    Blue:
      min: -76.0387  (-0.00116028)
      max: 6877.14 (0.104938)
      mean: 10.3803 (0.000158394)
      median: 6.83935 (0.000104362)
      standard deviation: 24.8323 (0.000378917)
      kurtosis: 24019.4
      skewness: 120.302
      entropy: 0.383766
  Image statistics:
    Overall:
      min: -97.9251  (-0.00149424)
      max: 7686.34 (0.117286)
      mean: 11.405 (0.00017403)
      median: 8.51849 (0.000129984)
      standard deviation: 26.0442 (0.000397408)
      kurtosis: 21820.5
      skewness: 113.133
      entropy: 0.380988
  Rendering intent: Perceptual
  Gamma: 0.454545
  Chromaticity:
    red primary: (0.64,0.33)
    green primary: (0.3,0.6)
    blue primary: (0.15,0.06)
    white point: (0.3127,0.329)
  Matte color: grey74
  Background color: white
  Border color: srgb(223,223,223)
  Transparent color: none
  Interlace: None
  Intensity: Undefined
  Compose: Over
  Page geometry: 5488x3664+0+0
  Dispose: Undefined
  Iterations: 0
  Compression: None
  Orientation: TopLeft
  Profiles:
    Profile-icc: 6900 bytes
  Properties:
    comment: SIMPLE  =                    T / file does conform to FITS standard
BITPIX  =                  -32 / number of bits per data pixel
NAXIS   =                    3 / number of data axes
NAXIS1  =                 5488 / length of data axis 1
NAXIS2  =                 3664 / length of data axis 2
NAXIS3  =                    3 / length of data axis 3
EXTEND  =                    T / FITS dataset may contain extensions
COMMENT   FITS (Flexible Image Transport System) format is defined in 'Astronomy
COMMENT   and Astrophysics', volume 376, page 359; bibcode: 2001A&A...376..359H
BZERO   =                    0 / offset data range to that of unsigned short
BSCALE  =                    1 / default scaling factor
DATE    = '2023-02-12T09:44:49' / UTC date that FITS file was created
DATE-OBS= '2023-01-30T00:53:59' / YYYY-MM-DDThh:mm:ss observation start, UT
INSTRUME= 'Pentax K-S2'        / instrument name
OBSERVER= '        '           / observer name
TELESCOP= '        '           / telescope used to acquire this image
ROWORDER= 'BOTTOM-UP'          / Order of the rows in image array
XPIXSZ  =                  4.3 / X pixel size microns
YPIXSZ  =                  4.3 / Y pixel size microns
XBINNING=                    1 / Camera binning mode
YBINNING=                    1 / Camera binning mode
FOCALLEN=                  180 / Camera focal length
CCD-TEMP=                    0 / CCD temp in C
EXPTIME =                10.04 / Exposure time [s]
APERTURE=                  5.6 / Aperture of the instrument
ISOSPEED=                 6400 / ISO camera setting
HISTORY Calibrated with a master dark
HISTORY Calibrated with a master flat, normalization of 0.005
HISTORY Cosmetic correction of 0 cold pixels and 0 hot pixels
END

    date:create: 2023-02-12T09:45:21+00:00
    date:modify: 2023-02-12T09:45:21+00:00
    date:timestamp: 2023-02-12T10:02:47+00:00
    icc:copyright: Public Domain
    icc:description: Siril built-in sRGB
    icc:manufacturer: Siril
    icc:model: sRGB
    quantum:format: floating-point
    signature: fb0e8d629f992d84bc579bf828887ef17763c5197f90943c69991d7b3f967b34
    tiff:alpha: unspecified
    tiff:copyright: Siril v1.1.0
    tiff:endian: lsb
    tiff:photometric: RGB
    tiff:rows-per-strip: 1
    tiff:software: siril v1.1.0
    tiff:timestamp: 2023:01:30 00:53:59
  Artifacts:
    verbose: true
  Tainted: False
  Filesize: 230.155MiB
  Number pixels: 20.108M
  Pixel cache type: Memory
  Pixels per second: 83.2666MP
  User time: 0.240u
  Elapsed time: 0:01.241
  Version: ImageMagick 7.1.0-48 Q16-HDRI x86_64 c1ca24765:20220910 https://imagemagick.org
```

**Comment by FAb7D** [comment]
`magick r_r_pp_light_00039.fit -combine -channel rgb -auto-level x.tiff`
However there is something very wrong with the red versus green and blue curve it seems to me.
There is also this [https://discuss.pixls.us/t/unable-to-view-fit-output-from-imagemagick-in-siril/26380](https://discuss.pixls.us/t/unable-to-view-fit-output-from-imagemagick-in-siril/26380)

**Reply by fmw42** [reply]
I found the same. I suspect you need specific stretches for each channel, not just the auto-level of each.  I could not figure that out from the meta data.  Perhaps the data needs to be treated as linear before the combine and then convert to sRGB afterwards?

**Reply by FAb7D** [reply]
I asked the question (FIT->TIF) also  here ( [discuss.pixls.us](https://discuss.pixls.us/t/from-fit-of-siril-to-tiff-with-imagemagick/35386/5) )today.
And a little while ago I tried this that seems to work enough, but not perfect:
` magick r_r_pp_light_00039.fit -combine -channel gb -level 0%,-20% -channel r -level 0%,100% x-magick.tiff`
but I also start to assume that the curve R is actually G and vice versa.

**Comment by fmw42** [comment]
You will have to find the meta data that is used by FITs tools to do the scaling. I do not see what is needed in the information IM reports in magick identify -verbose. Somewhere in the meta data or documentation for FITS is that information. We cannot just guess.

**Reply by FAb7D** [reply]
Ok, thanks!

**Reply by fmw42** [reply]
Perhaps one of the IM developer might be able to find that in the FITs documentation. They will need to comment further.

**Reply by FAb7D** [reply]
I hope so and I'm not in a hurry, I'll wait.
